#### Exercise -1.01

Import the pandas and numpy packages using shorthand notation:


In [4]:
import pandas as pd
import numpy as np

Read .csv file using pandas and show the first 10 samples from the dataset

In [5]:
!ls

Exercises -1.01 to 1.05.ipynb titanic-Copy1.csv


In [6]:
data=pd.read_csv('titanic-Copy1.csv')

In [7]:
data.head(10)

Unnamed: 0 Cabin Embarked     Fare  Pclass            Ticket   Age  \
0           0   NaN        S   7.2500       3         A/5 21171  22.0   
1           1   C85        C  71.2833       1          PC 17599  38.0   
2           2   NaN        S   7.9250       3  STON/O2. 3101282  26.0   
3           3  C123        S  53.1000       1            113803  35.0   
4           4   NaN        S   8.0500       3            373450  35.0   
5           5   NaN        Q   8.4583       3            330877   NaN   
6           6   E46        S  51.8625       1             17463  54.0   
7           7   NaN        S  21.0750       3            349909   2.0   
8           8   NaN        S  11.1333       3            347742  27.0   
9           9   NaN        C  30.0708       2            237736  14.0   

                                                Name  Parch     Sex  SibSp  \
0                            Braund, Mr. Owen Harris      0    male      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...      0  female      1   
2                             Heikkinen, Miss. Laina      0  female      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)      0  female      1   
4                           Allen, Mr. William Henry      0    male      0   
5                                   Moran, Mr. James      0    male      0   
6                            McCarthy, Mr. Timothy J      0    male      0   
7                     Palsson, Master. Gosta Leonard      1    male      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)      2  female      0   
9                Nasser, Mrs. Nicholas (Adele Achem)      0  female      1   

   Survived  
0       0.0  
1       1.0  
2       1.0  
3       1.0  
4       0.0  
5       0.0  
6       0.0  
7       0.0  
8       1.0  
9       1.0

#### Exercise-1.02

Select individual columns in a similar way to a regular dictionary by using the labels of the columns, as shown here:


In [8]:
data['Age']

0       22.0
1       38.0
2       26.0
3       35.0
4       35.0
        ... 
1304     NaN
1305    39.0
1306    38.5
1307     NaN
1308     NaN
Name: Age, Length: 1309, dtype: float64

If there are no spaces in the column name, we can also use the dot operator. If there are spaces in the column names, we will need to use the bracket notation:


In [9]:
data.Age

0       22.0
1       38.0
2       26.0
3       35.0
4       35.0
        ... 
1304     NaN
1305    39.0
1306    38.5
1307     NaN
1308     NaN
Name: Age, Length: 1309, dtype: float64

Select multiple columns at once using bracket notation, as shown here:


In [10]:
data[['Cabin','Age','Embarked']]

Cabin   Age Embarked
0      NaN  22.0        S
1      C85  38.0        C
2      NaN  26.0        S
3     C123  35.0        S
4      NaN  35.0        S
...    ...   ...      ...
1304   NaN   NaN        S
1305  C105  39.0        C
1306   NaN  38.5        S
1307   NaN   NaN        S
1308   NaN   NaN        C

[1309 rows x 3 columns]

Select the first row using iloc:


In [11]:
data.iloc[0]

Unnamed: 0                          0
Cabin                             NaN
Embarked                            S
Fare                             7.25
Pclass                              3
Ticket                      A/5 21171
Age                                22
Name          Braund, Mr. Owen Harris
Parch                               0
Sex                              male
SibSp                               1
Survived                            0
Name: 0, dtype: object

Select the first three rows using iloc:


In [12]:
data.iloc[[0,1,2]]

Unnamed: 0 Cabin Embarked     Fare  Pclass            Ticket   Age  \
0           0   NaN        S   7.2500       3         A/5 21171  22.0   
1           1   C85        C  71.2833       1          PC 17599  38.0   
2           2   NaN        S   7.9250       3  STON/O2. 3101282  26.0   

                                                Name  Parch     Sex  SibSp  \
0                            Braund, Mr. Owen Harris      0    male      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...      0  female      1   
2                             Heikkinen, Miss. Laina      0  female      0   

   Survived  
0       0.0  
1       1.0  
2       1.0

In [13]:
data.iloc[0:3]

Unnamed: 0 Cabin Embarked     Fare  Pclass            Ticket   Age  \
0           0   NaN        S   7.2500       3         A/5 21171  22.0   
1           1   C85        C  71.2833       1          PC 17599  38.0   
2           2   NaN        S   7.9250       3  STON/O2. 3101282  26.0   

                                                Name  Parch     Sex  SibSp  \
0                            Braund, Mr. Owen Harris      0    male      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...      0  female      1   
2                             Heikkinen, Miss. Laina      0  female      0   

   Survived  
0       0.0  
1       1.0  
2       1.0

Next, get a list of all of the available columns:


In [14]:
cols = data.columns
cols

Index(['Unnamed: 0', 'Cabin', 'Embarked', 'Fare', 'Pclass', 'Ticket', 'Age',
       'Name', 'Parch', 'Sex', 'SibSp', 'Survived'],
      dtype='object')

Use this list of columns and the standard Python slicing syntax to get columns 2, 3, and 4, and their corresponding values:


In [15]:
data[cols[1:4]]

Cabin Embarked      Fare
0      NaN        S    7.2500
1      C85        C   71.2833
2      NaN        S    7.9250
3     C123        S   53.1000
4      NaN        S    8.0500
...    ...      ...       ...
1304   NaN        S    8.0500
1305  C105        C  108.9000
1306   NaN        S    7.2500
1307   NaN        S    8.0500
1308   NaN        C   22.3583

[1309 rows x 3 columns]

Use the len operator to get the number of rows in the DataFrame:


In [16]:
len(data)

1309

Get the value for the Fare column in row 2 using the row-centric method:


In [17]:
data.iloc[2]['Fare']

7.925

Use the dot operator for the column, as follows:


In [18]:
data['Fare'].iloc[2]

7.925

Use the column-centric method, as follows:


In [19]:
data['Fare'][2]

7.925

Use the column-centric method with the dot operator, as follows:


In [20]:
data.Fare[2]

7.925

#### Exercise 1.03


Create a list of the passengers' names and ages for those passengers under the age of 21, as shown here:


In [21]:
child_passenger = data[data.Age <21][['Name','Age']]
child_passenger

Name   Age
7                     Palsson, Master. Gosta Leonard   2.0
9                Nasser, Mrs. Nicholas (Adele Achem)  14.0
10                   Sandstrom, Miss. Marguerite Rut   4.0
12                    Saundercock, Mr. William Henry  20.0
14              Vestrom, Miss. Hulda Amanda Adolfina  14.0
...                                              ...   ...
1283                   Abbott, Master. Eugene Joseph  13.0
1286  Smith, Mrs. Lucien Philip (Mary Eloise Hughes)  18.0
1294                          Carrau, Mr. Jose Pedro  17.0
1296    Nourney, Mr. Alfred (Baron von Drachstedt")"  20.0
1300                       Peacock, Miss. Treasteall   3.0

[249 rows x 2 columns]

In [22]:
data.loc[data.Age <21][['Name','Age']]

Name   Age
7                     Palsson, Master. Gosta Leonard   2.0
9                Nasser, Mrs. Nicholas (Adele Achem)  14.0
10                   Sandstrom, Miss. Marguerite Rut   4.0
12                    Saundercock, Mr. William Henry  20.0
14              Vestrom, Miss. Hulda Amanda Adolfina  14.0
...                                              ...   ...
1283                   Abbott, Master. Eugene Joseph  13.0
1286  Smith, Mrs. Lucien Philip (Mary Eloise Hughes)  18.0
1294                          Carrau, Mr. Jose Pedro  17.0
1296    Nourney, Mr. Alfred (Baron von Drachstedt")"  20.0
1300                       Peacock, Miss. Treasteall   3.0

[249 rows x 2 columns]

Count how many child passengers there were, as shown here:


In [23]:
len(child_passenger)

249

Count how many passengers were between the ages of 21 and 30. Do not use Python's and logical operator for this step, but rather the ampersand symbol (&). Do this as follows:


In [24]:
data.loc[(data.Age>21) & (data.Age<30)]

Unnamed: 0 Cabin Embarked     Fare  Pclass            Ticket   Age  \
0              0   NaN        S   7.2500       3         A/5 21171  22.0   
2              2   NaN        S   7.9250       3  STON/O2. 3101282  26.0   
8              8   NaN        S  11.1333       3            347742  27.0   
23            23    A6        S  35.5000       1            113788  28.0   
34            34   NaN        C  82.1708       1          PC 17604  28.0   
...          ...   ...      ...      ...     ...               ...   ...   
1287        1287   NaN        Q   7.2500       3            371109  24.0   
1289        1289   NaN        S   7.7750       3            347065  22.0   
1293        1293   NaN        C  59.4000       1            112378  22.0   
1297        1297   NaN        S  10.5000       2             28666  23.0   
1303        1303   NaN        S   7.7750       3            347086  28.0   

                                                   Name  Parch     Sex  SibSp  \
0                               Braund, Mr. Owen Harris      0    male      1   
2                                Heikkinen, Miss. Laina      0  female      0   
8     Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)      2  female      0   
23                         Sloper, Mr. William Thompson      0    male      0   
34                              Meyer, Mr. Edgar Joseph      0    male      1   
...                                                 ...    ...     ...    ...   
1287                               Colbert, Mr. Patrick      0    male      0   
1289                     Larsson-Rondberg, Mr. Edvard A      0    male      0   
1293                     Gibson, Miss. Dorothy Winifred      1  female      0   
1297                          Ware, Mr. William Jeffery      0    male      1   
1303                     Henriksson, Miss. Jenny Lovisa      0  female      0   

      Survived  
0          0.0  
2          1.0  
8          1.0  
23         1.0  
34         0.0  
...        ...  
1287       NaN  
1289       NaN  
1293       NaN  
1297       NaN  
1303       NaN  

[279 rows x 12 columns]

Find the passengers who were either first- or third-class ticket holders. Again, we will not use the Python logical or operator but the pipe symbol (|) instead. Do this as follows:


In [25]:
data.loc[(data.Pclass ==1) |(data.Pclass == 3)]

Unnamed: 0 Cabin Embarked      Fare  Pclass              Ticket   Age  \
0              0   NaN        S    7.2500       3           A/5 21171  22.0   
1              1   C85        C   71.2833       1            PC 17599  38.0   
2              2   NaN        S    7.9250       3    STON/O2. 3101282  26.0   
3              3  C123        S   53.1000       1              113803  35.0   
4              4   NaN        S    8.0500       3              373450  35.0   
...          ...   ...      ...       ...     ...                 ...   ...   
1304        1304   NaN        S    8.0500       3           A.5. 3236   NaN   
1305        1305  C105        C  108.9000       1            PC 17758  39.0   
1306        1306   NaN        S    7.2500       3  SOTON/O.Q. 3101262  38.5   
1307        1307   NaN        S    8.0500       3              359309   NaN   
1308        1308   NaN        C   22.3583       3                2668   NaN   

                                                   Name  Parch     Sex  SibSp  \
0                               Braund, Mr. Owen Harris      0    male      1   
1     Cumings, Mrs. John Bradley (Florence Briggs Th...      0  female      1   
2                                Heikkinen, Miss. Laina      0  female      0   
3          Futrelle, Mrs. Jacques Heath (Lily May Peel)      0  female      1   
4                              Allen, Mr. William Henry      0    male      0   
...                                                 ...    ...     ...    ...   
1304                                 Spector, Mr. Woolf      0    male      0   
1305                       Oliva y Ocana, Dona. Fermina      0  female      0   
1306                       Saether, Mr. Simon Sivertsen      0    male      0   
1307                                Ware, Mr. Frederick      0    male      0   
1308                           Peter, Master. Michael J      1    male      1   

      Survived  
0          0.0  
1          1.0  
2          1.0  
3          1.0  
4          0.0  
...        ...  
1304       NaN  
1305       NaN  
1306       NaN  
1307       NaN  
1308       NaN  

[1032 rows x 12 columns]

Find the passengers who were not holders of either first- or third-class tickets. Do not simply select those second-class ticket holders, but use the ~ symbol for the not logical operator instead. Do this as follows:


In [26]:
data.loc[~((data.Pclass == 1) |(data.Pclass==3))]

Unnamed: 0 Cabin Embarked     Fare  Pclass         Ticket   Age  \
9              9   NaN        C  30.0708       2         237736  14.0   
15            15   NaN        S  16.0000       2         248706  55.0   
17            17   NaN        S  13.0000       2         244373   NaN   
20            20   NaN        S  26.0000       2         239865  35.0   
21            21   D56        S  13.0000       2         248698  34.0   
...          ...   ...      ...      ...     ...            ...   ...   
1278        1278   NaN        S  13.0000       2         244346  57.0   
1284        1284   NaN        S  10.5000       2     C.A. 30769  47.0   
1292        1292   NaN        S  21.0000       2          28664  38.0   
1296        1296   D38        C  13.8625       2  SC/PARIS 2166  20.0   
1297        1297   NaN        S  10.5000       2          28666  23.0   

                                              Name  Parch     Sex  SibSp  \
9              Nasser, Mrs. Nicholas (Adele Achem)      0  female      1   
15                Hewlett, Mrs. (Mary D Kingcome)       0  female      0   
17                    Williams, Mr. Charles Eugene      0    male      0   
20                            Fynney, Mr. Joseph J      0    male      0   
21                           Beesley, Mr. Lawrence      0    male      0   
...                                            ...    ...     ...    ...   
1278                               Ashby, Mr. John      0    male      0   
1284                          Gilbert, Mr. William      0    male      0   
1292                               Gale, Mr. Harry      0    male      1   
1296  Nourney, Mr. Alfred (Baron von Drachstedt")"      0    male      0   
1297                     Ware, Mr. William Jeffery      0    male      1   

      Survived  
9          1.0  
15         1.0  
17         1.0  
20         0.0  
21         1.0  
...        ...  
1278       NaN  
1284       NaN  
1292       NaN  
1296       NaN  
1297       NaN  

[277 rows x 12 columns]

We no longer need the Unnamed: 0 column, so delete it using the del operator:


In [27]:
del data['Unnamed: 0']

In [28]:
data.head()

Cabin Embarked     Fare  Pclass            Ticket   Age  \
0   NaN        S   7.2500       3         A/5 21171  22.0   
1   C85        C  71.2833       1          PC 17599  38.0   
2   NaN        S   7.9250       3  STON/O2. 3101282  26.0   
3  C123        S  53.1000       1            113803  35.0   
4   NaN        S   8.0500       3            373450  35.0   

                                                Name  Parch     Sex  SibSp  \
0                            Braund, Mr. Owen Harris      0    male      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...      0  female      1   
2                             Heikkinen, Miss. Laina      0  female      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)      0  female      1   
4                           Allen, Mr. William Henry      0    male      0   

   Survived  
0       0.0  
1       1.0  
2       1.0  
3       1.0  
4       0.0

#### Exercise-1.04

Use the groupby method to group the data under the Embarked column to find out how many different values for Embarked there are:


In [30]:
embarked_group = data.groupby('Embarked')

In [33]:
len(embarked_group)

3

Display the output of embarked_grouped.groups to find what the groupby method actually does:


In [31]:
embarked_group.groups

{'C': Int64Index([   1,    9,   19,   26,   30,   31,   34,   36,   39,   42,
             ...
             1260, 1262, 1266, 1288, 1293, 1295, 1296, 1298, 1305, 1308],
            dtype='int64', length=270),
 'Q': Int64Index([   5,   16,   22,   28,   32,   44,   46,   47,   82,  109,
             ...
             1206, 1249, 1271, 1272, 1279, 1287, 1290, 1299, 1301, 1302],
            dtype='int64', length=123),
 'S': Int64Index([   0,    2,    3,    4,    6,    7,    8,   10,   11,   12,
             ...
             1289, 1291, 1292, 1294, 1297, 1300, 1303, 1304, 1306, 1307],
            dtype='int64', length=914)}

In [40]:
embarked_group

Use the iloc method to inspect row 1 and confirm that it belongs to embarked group C:


In [34]:
data.iloc[1]

Cabin                                                     C85
Embarked                                                    C
Fare                                                  71.2833
Pclass                                                      1
Ticket                                               PC 17599
Age                                                        38
Name        Cumings, Mrs. John Bradley (Florence Briggs Th...
Parch                                                       0
Sex                                                    female
SibSp                                                       1
Survived                                                    1
Name: 1, dtype: object

As the groups are a dictionary, we can iterate through them and execute computations on the individual groups. Compute the mean age for each group, as shown here:


In [49]:
for name,group in embarked_group:
    print(f'name of the group: {name}\t mean age: {np.mean(group.Age)}\n')

name of the group: C	 mean age: 32.33216981132075

name of the group: Q	 mean age: 28.63

name of the group: S	 mean age: 29.245204603580564



In [46]:
for name,group in embarked_group:
    print(f'name of the group: {name} top samples:{group.head(2)}\n')
    

name of the group: C top samples:  Cabin Embarked     Fare  Pclass    Ticket   Age  \
1   C85        C  71.2833       1  PC 17599  38.0   
9   NaN        C  30.0708       2    237736  14.0   

                                                Name  Parch     Sex  SibSp  \
1  Cumings, Mrs. John Bradley (Florence Briggs Th...      0  female      1   
9                Nasser, Mrs. Nicholas (Adele Achem)      0  female      1   

   Survived  
1       1.0  
9       1.0  

name of the group: Q top samples:   Cabin Embarked     Fare  Pclass  Ticket  Age                  Name  Parch  \
5    NaN        Q   8.4583       3  330877  NaN      Moran, Mr. James      0   
16   NaN        Q  29.1250       3  382652  2.0  Rice, Master. Eugene      1   

     Sex  SibSp  Survived  
5   male      0       0.0  
16  male      4       0.0  

name of the group: S top samples:  Cabin Embarked   Fare  Pclass            Ticket   Age  \
0   NaN        S  7.250       3         A/5 21171  22.0   
2   NaN        S  7

Another option is to use the aggregate method, or agg for short, and provide it with the function to apply across the columns. Use the agg method to determine the mean of each group:


In [50]:
embarked_group.agg(np.mean)

Fare    Pclass        Age     Parch     SibSp  Survived
Embarked                                                              
C         62.336267  1.851852  32.332170  0.370370  0.400000  0.553571
Q         12.409012  2.894309  28.630000  0.113821  0.341463  0.389610
S         27.418824  2.347921  29.245205  0.426696  0.550328  0.336957

#### Exercise 1.05

Write the first_val function as a Lambda function, passed to agg:


In [51]:
embarked_group.agg(lambda x : x.values[0])

Cabin     Fare  Pclass     Ticket   Age  \
Embarked                                           
C          C85  71.2833       1   PC 17599  38.0   
Q          NaN   8.4583       3     330877   NaN   
S          NaN   7.2500       3  A/5 21171  22.0   

                                                       Name  Parch     Sex  \
Embarked                                                                     
C         Cumings, Mrs. John Bradley (Florence Briggs Th...      0  female   
Q                                          Moran, Mr. James      0    male   
S                                   Braund, Mr. Owen Harris      0    male   

          SibSp  Survived  
Embarked                   
C             1       1.0  
Q             0       0.0  
S             1       0.0

We can also pass multiple functions to agg via a list to apply the functions across the dataset. Pass the Lambda function as well as the NumPy mean and standard deviation functions, like this:


In [52]:
embarked_group.agg([lambda x:x.values[0], np.mean,np.std])

Fare                           Pclass                      \
         <lambda_0>       mean        std <lambda_0>      mean       std   
Embarked                                                                   
C           71.2833  62.336267  84.185996          1  1.851852  0.936802   
Q            8.4583  12.409012  13.616133          3  2.894309  0.380099   
S            7.2500  27.418824  37.096402          3  2.347921  0.784126   

                Age                            Parch                      \
         <lambda_0>       mean        std <lambda_0>      mean       std   
Embarked                                                                   
C              38.0  32.332170  15.258092          0  0.370370  0.670579   
Q               NaN  28.630000  15.045784          0  0.113821  0.531056   
S              22.0  29.245205  14.047507          0  0.426696  0.943989   

              SibSp                       Survived                      
         <lambda_0>      mean       std <lambda_0>      mean       std  
Embarked                                                                
C                 1  0.400000  0.554803        1.0  0.553571  0.498608  
Q                 0  0.341463  0.885487        0.0  0.389610  0.490860  
S                 1  0.550328  1.161723        0.0  0.336957  0.473037

Apply numpy.sum to the Fare column and the Lambda function to the Age column by passing agg a dictionary where the keys are the columns to apply the function to, and the values are the functions themselves to be able to apply different functions to different columns in the DataFrame:


In [55]:
embarked_group.agg({'Fare':np.sum, 
                    'Age':lambda x:x.values[0]})

Fare   Age
Embarked                  
C         16830.7922  38.0
Q          1526.3085   NaN
S         25033.3862  22.0

Finally, execute the groupby method using more than one column. Provide the method with a list of the columns (Sex and Embarked) to groupby, like this:


In [57]:
double_grp =data.groupby(['Embarked','Age'])
double_grp.groups

{('S',
  22.0): Int64Index([   0,   80,  112,  141,  151,  212,  225,  243,  287,  320,  323,
              356,  376,  395,  474,  478,  521,  554,  588,  708,  882,  895,
              981, 1017, 1038, 1060, 1113, 1137, 1170, 1210, 1258, 1267, 1289],
            dtype='int64'),
 ('C', 38.0): Int64Index([1, 716, 1238], dtype='int64'),
 ('S',
  26.0): Int64Index([   2,   69,   93,  162,  290,  312,  315,  401,  509,  617,  619,
              628,  704,  810,  870,  898,  994, 1028, 1050, 1053, 1056, 1186,
             1211, 1213, 1237],
            dtype='int64'),
 ('S',
  35.0): Int64Index([3, 4, 20, 211, 230, 269, 279, 363, 383, 486, 590, 614, 701, 812,
             947],
            dtype='int64'),
 ('Q', nan): Int64Index([5], dtype='int64'),
 ('S', 54.0): Int64Index([6, 124, 249, 317, 582, 774, 1265], dtype='int64'),
 ('S',
  2.0): Int64Index([7, 119, 205, 297, 340, 479, 530, 642, 824, 980, 1175], dtype='int64'),
 ('S',
  27.0): Int64Index([   8,   41,  146,  216,  221,  321,  551,